### DISCUSSION TITLE
Form values change with state updates?

### DISCUSSION THREAD

**Discussion by naomei** [react-hook-form/react-hook-form#6888] [Q&A] [answered]
As a summary, changing a state that has nothing to do with the form will change the form values.

## code-sandbox

https://codesandbox.io/s/react-hook-form-test-gvmvn

### 1. form setting
```ts
const form = useForm<{ user?: { name?: string } }>({
  defaultValues: { user: { name: "default-name" } }
});
```

### 2. click "clear" button
Change user to `undefined`.

### 3. click "change-count" button
Change local-state.

### 4. click "submit" and see console
I expect `{user: undefined}`, but the result is `{user: {name: "default-name"}}`.

---

Please tell me if this is the expected behavior.

**Comment by bluebill1049** [comment, marked as answer]
nice and detailed question! we don't support `undefned` for the input value.

> The value for the field. Make sure you supply the entire array when you update useFieldArray and valuecan't be undefined.

## reason

-  native input doesn't support `undefined` (we do support uncontrolled input)
- `undefined` is perceived value for us to detect whether use inline value (defaultValue) or defaultValues from useForm

**Reply by naomei** [reply]
@bluebill1049 

Thank you for your polite answer.

> native input doesn't support

oh, I completely forgot...

> undefined is perceived value for us to detect whether use inline value (defaultValue) or defaultValues from useForm

Thank you, I understand roughly!

**Reply by bluebill1049** [reply]
let me provide some more context:

- e.target.value = undefined // ❌
- `<Controller />` // defaultValue undefined so we look up defaultValues
- `watch()` // defaultValue  undefined so we look up defaultValues

**Reply by naomei** [reply]
@bluebill1049 

Thank you for the additional note! I fully understood.

**Reply by naomei** [reply]
@bluebill1049 

I'm sorry, let me have a little more conversation...

I understand the behavior when `undefined` .
How about `null` ?
(I know that native `input.value` doesn't support `null` as well as `undefined`)

## Example 
Suppose want to manage the following data as a form.

```ts
type FormValue = {
   name: string;
   job: { type: string }
}
```

There are cases where the job value should be temporarily set to an "invalid value" by operating the UI.
What should I do in that case?

**Reply by bluebill1049** [reply]
null is `valid` value, although if you want to be as close to the native input, then an empty string is probably the preferred default value.  Personally, I prefer to stay closer to the native standard for an input value, and not apply too much transformation with the value.
